addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SinkDetectionVerifyProbableSinks.java

```diff
@@ -0,0 +1,142 @@
+
+    private static Logger log = Logger.getLogger(SinkDetectionVerifyProbableSinks.class);
+
+    @Override
+    public int getId() {
+        return 40037;
+    }
+
+    @Override
+    public String getName() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "name");
+    }
+
+    @Override
+    public String getDescription() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public int getCategory() {
+        return Category.INJECTION;
+    }
+
+    @Override
+    public String[] getDependency() {
+        return new String[] {"SinkDetectionFindProbableSinks"};
+    }
+
+    @Override
+    public String getSolution() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public String getReference() {
+        return Constant.messages.getString(MESSAGE_PREFIX + "misc");
+    }
+
+    @Override
+    public void scan(HttpMessage msg, String param, String value) {
+        // TODO: should test only on POST/PUT ?
+        SinkDetectionStorage storage = getStorage();
+        List<HttpMessage> possibleSinks = storage.getPossibleSinksForValue(value);
+
+        try {
+            HttpMessage inputMsg = msg.cloneRequest();
+            String injectedValue = getRandomString();
+            this.setParameter(inputMsg, param, injectedValue);
+            sendAndReceive(inputMsg, false);
+
+            for (HttpMessage possibleSink : possibleSinks) {
+                // Reflections on the same request are not considered "stored sinks"
+                if (inputMsg.hashCode() != possibleSink.hashCode()) {
+                    HttpMessage possibleSinkClone = possibleSink.cloneRequest();
+                    sendAndReceive(possibleSinkClone, false);
+                    if (possibleSinkClone.getResponseBody().toString().contains(injectedValue)) {
+                        // the sink source needs to be the original message otherwise path
+                        // params will not be stored properly
+                        ParamSinksUtils.setSinkForSource(msg, param, possibleSinkClone);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            log.error(e.getMessage(), e);
+        }
+    }
+
+    private SinkDetectionStorage getStorage() {
+        SinkDetectionStorage storage;
+        Object obj = getKb().get(SINK_DETECTION_STORAGE);
+        if (obj instanceof SinkDetectionStorage) {
+            storage = (SinkDetectionStorage) obj;
+        } else {
+            throw new IllegalStateException(
+                    "The SINK_DETECTION_STORAGE Should have been initialized by the SinkDetectionCollectAndRefreshParamValues");
+        }
+        return storage;
+    }
+
+    private static String getRandomString() {
+        String alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+        Random rand = new Random();
```

Comment:
Please check random works as expected with default constructor (No same seed is used) or we need to give a seed as timestamp to it.